### DISCUSSION TITLE
Is it possible to connect to a device from the onResult (Scan) callback?

### DISCUSSION THREAD

**Discussion by sivar2311** [h2zero/NimBLE-Arduino#373] [Q&A] [answered]
In my current project I would like to establish a connection to a device via the onResult callback of a NimBLEScan as soon as the device I am looking for is detected (asynchronous). Unfortunately, this does not seem to be possible - the connection is not established. 

If I connect from outside the callback (by calling a function in the main loop like `myClass.handleDevices()`) everything works as expected (synchronous). But this requires an extra call in the main loop, which I would like to avoid. 

Do you see a way to avoid this extra call? 
This would make my entire project work asynchronously.

**Comment by h2zero** [comment]
On an advanced level yes this is possible (direct calls to the NimBLE stack), however, from this library API it is not, due to task blocking and other things. There are other ways though, such as having the result callback unblock a task that then performs the connection, which is my method currently.

**Reply by sivar2311** [reply]
I wanted to avoid direct calls to stack, but if it helps I will not avoid it. Can you give me some tips in the right direction here about what calls are involved?

I suspect the second method is more appropriate. Do you have a little code snippet for me to point me in the right direction?

**Comment by h2zero** [comment, marked as answer]
>I wanted to avoid direct calls to stack, but if it helps I will not avoid it. Can you give me some tips in the right direction here about what calls are involved?

This would be exceptionally difficult to make work with the rest of the library, you would need to write your own client code and callback handling etc.. I certainly would not suggest it, just being thorough with my answer since it is possible. 

 >I suspect the second method is more appropriate. Do you have a little code snippet for me to point me in the right direction?

If you're familiar with FreeRTOS queues you could just have the connection handler task block on a queue receive call and push the device pointer into the queue from the scan callback (some priority management and task stack sizing my be required here) but that's the gist of it. There are other ways as well but this method is straight forward. I don't have any basic snippets at the ready right now but I'll see if I can give an example tomorrow.

**Reply by sivar2311** [reply]
I only have basic knowledge of FreeRTOS - but I am willing to acquire the necessary knowledge. 
I would be very happy about a small example - thank you in advance and have a good night! 
(Here it is just 6:15 in the morning)

**Comment by sivar2311** [comment]
I think I have found a suitable solution with `xTaskCreate` to run the connection code in a separate task (without the need for a TaskQueue):
```C++
void MyClass::begin() {
        ...
        xTaskCreate(connect_task, "connect", 2048, this, tskIDLE_PRIORITY, nullptr);
}

void MyClass::connect_task(void* vParams) {
    MyClass* instance = (MyClass*)vParams;

    for (;;) {
        if (instance->found_connectable_device()) {
            NimBLEAdvertisedDevice* device = instance->get_connectable_device();
            instance->connect_device(device);
        }
        vTaskDelay(1); // feed the watchdog
    }
}
```

**Reply by h2zero** [reply]
Yes, that is another way to do it 👍. You might want to bump up the priority though, tskIDLE_PRIORITY+1 would be more responsive.

**Reply by h2zero** [reply]
Here is an example with using queues, which are preferable so the task isn't always running.

https://github.com/1technophile/OpenMQTTGateway/blob/fb56162c52a65fbfa82e227fefb1b187e4187327/main/ZgatewayBT.ino#L385

**Reply by sivar2311** [reply]
This is the first time I have used xTaskCreate. Thanks for the info about task priority. 
I will go through the linked code and try to find out the relevant parts for me.
Thanks a lot!

**Reply by sivar2311** [reply]
I understood the concept with the queue. Unfortunately, it's not quite as easy for me to implement. This is probably because I don't use the cache for the advertising devices. But I already have a working solution thanks to your tips :)

I will probably have to deal more with FreeRTOS in the future.
